### PR TITLE
ASA 8404 (#218)

### PR DESCRIPTION
**Pull request by vishalhcl-5960**
Include-SCA implementation

### CODE REVIEW

**Review by mattmurp — COMMENTED** [review]
I understand that a lot of the changes here were made in order to run 2 scans at once (i.e. call the perform method twice). However, there is a lot of if/else/if/else logic that's hard to follow and there are a lot of classes (e.g. AppScanBuildStep, ResultsRetriever) that now have specific knowledge of different types of scans.
One solution to these problems would be to create a new type of scan that will run both a SAST and SCA scan.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java [hunk 1]
```diff
@@ -23,15 +23,29 @@ public abstract class Scanner extends AbstractDescribableImpl<Scanner> implement
...
 	}
 	
 	public boolean getHasOptions() {
 		return m_hasOptions;
 	}
+
+	public boolean hasOptions() {
+		return m_hasOptions;
+	}
+
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java]: Okay sure, added the "hasOptionsUploadDirect()" methods in the StaticAnalyzer class.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java [hunk 2]
```diff
@@ -23,15 +23,29 @@ public abstract class Scanner extends AbstractDescribableImpl<Scanner> implement
...
 		m_target = target;
 		m_hasOptions = hasOptions;
+		m_hasOptionsUploadDirect = hasOptionsUploadDirect;
 	}
 	
 	public boolean getHasOptions() {
 		return m_hasOptions;
 	}
+
+	public boolean hasOptions() {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java]: "hasOptions" was used as a checker method in the jelly files. Now, using the getHasOptions() itself in the jelly files too.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java [hunk 3]
```diff
@@ -23,15 +23,29 @@ public abstract class Scanner extends AbstractDescribableImpl<Scanner> implement
 	
 	private String m_target;
 	private boolean m_hasOptions;
+	private boolean m_hasOptionsUploadDirect;
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java]: Moving this property to the StaticAnalyzer class.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java [hunk 4]
```diff
@@ -328,89 +364,122 @@ private void validations(boolean isAppScan360, Map<String, String> properties, I
             throw new AbortException(Messages.error_url_validation(target));
         }
     }
+
+    private String updatedScanTypesForBuildSummaryName(String scanType) {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java:368]: Added this method in the Scanner class from which i am calling that method in the getProperties() method of respective scan classes.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java [hunk 5]
```diff
@@ -235,11 +236,33 @@ public boolean getPersonalScan() {
     public DescriptorImpl getDescriptor() {
         return (DescriptorImpl)super.getDescriptor();
     }
+
+    //to execute SAST & SCA scans consecutively, 1st perform method call to execute SAST then after changing the scan type to SCA
+    private void includeSCAImplementation(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java:241]: Moved the validations which are returning error message in the respective scan classes by adding an extra argument in the getProperties() method.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java [hunk 6]
```diff
@@ -309,6 +338,8 @@ private void validations(boolean isAppScan360, Map<String, String> properties, I
         if(isAppScan360) {
             if (m_type.equals(Scanner.DYNAMIC_ANALYZER) && properties.containsKey(Scanner.PRESENCE_ID)) {
                 throw new AbortException(Messages.error_presence_AppScan360());
+            } if(properties.containsKey(CoreConstants.INCLUDE_SCA)) {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java:341]: Moved the validations in the respective scan classes by adding an "validations()" method in the abstract class.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java [hunk 7]
```diff
@@ -28,31 +28,40 @@ public class StaticAnalyzer extends Scanner {
...
 	@DataBoundConstructor
-	public StaticAnalyzer(String target,boolean hasOptions) {
+	public StaticAnalyzer(String target,boolean hasOptions, boolean hasOptionsUploadDirect) {
 		super(target, hasOptions);
                 m_openSourceOnly=false;
                 m_sourceCodeOnly=false;
                 m_scanMethod=CoreConstants.CREATE_IRX;
                 m_scanSpeed="";
-	}
+                m_includeSCAGenerateIRX="true";
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:61]: I think it is better to have two different names for the these checkbox as the position of these checks are different in jelly file, it will hamper the value of the checkbox if one of the check is false.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java:368]: I can define this method in the Utils class of the SDK.

File: src/main/java/com/hcl/appscan/jenkins/plugin/results/ResultsInspector.java [hunk 8]
```diff
@@ -48,4 +61,55 @@ private boolean exceedsThreshold(String type, int threshold) {
 			return false;
 		}
 	}
+
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/results/ResultsInspector.java:64]: I have created a new class "CombinedResultsInspector" which is being extended from the base "ResultsInspector" class.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java [hunk 9]
```diff
@@ -235,15 +237,34 @@ public boolean getPersonalScan() {
     public DescriptorImpl getDescriptor() {
         return (DescriptorImpl)super.getDescriptor();
     }
-    
-    @Override
-    public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
-	perform((Run<?,?>)build, launcher, listener);
-	        return true;
+
+    //to execute SAST & SCA scans consecutively, 1st perform method call to execute SAST then after changing the scan type to SCA
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java:241]: Rather than having all of these checks for SCA and SAST and rather than trying to execute the perform() method twice, we should create a new scanner type in the SDK that will execute both a SAST scan and an SCA scan. That would eliminate the need for most, if not all, of these changes.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java [hunk 10]
```diff
@@ -38,7 +40,9 @@ public String getTarget() {
 	}
 	
 	public abstract Map<String, String> getProperties(VariableResolver<String> resolver) throws AbortException;
-	
+
+	public abstract void validations(IAuthenticationProvider authProvider, Map<String, String> properties, IProgress progress) throws AbortException;
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/Scanner.java:44]: In DynamicAnalyzer class we are doing the URL validity check & for that we need to call the getProxy() method from authProvider class. Instead of passing the IAuthenticationProvider, I will pass the JenkinsAuthenaticationProvider itself.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/DynamicAnalyzer.java [hunk 11]
```diff
@@ -196,8 +198,18 @@ public String upgradeLoginScenario(){
 		}
 	}
 
+	public void validations(IAuthenticationProvider authProvider, Map<String, String> properties, IProgress progress) throws AbortException {
+		if (((JenkinsAuthenticationProvider) authProvider).isAppScan360()) {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/DynamicAnalyzer.java:202]: okay, sure.

File: src/main/java/com/hcl/appscan/jenkins/plugin/results/ResultsInspector.java [hunk 12]
```diff
@@ -10,14 +10,13 @@
 
 import com.hcl.appscan.sdk.results.IResultsProvider;
 
-public class ResultsInspector {
+public class ResultsInspector extends CombinedResultsInspector {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/results/ResultsInspector.java]: Removed the CombinedResultsInspector class as we are now handling the results in the SDK.

File: src/main/java/com/hcl/appscan/jenkins/plugin/results/CombinedResultsInspector.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
+package com.hcl.appscan.jenkins.plugin.results;
+
+import com.hcl.appscan.jenkins.plugin.scanners.Scanner;
+import com.hcl.appscan.sdk.CoreConstants;
+import com.hcl.appscan.sdk.results.IResultsProvider;
+
+import java.util.List;
+import java.util.Map;
+
+public class CombinedResultsInspector {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/results/CombinedResultsInspector.java]: Removed the CombinedResultsInspector class as we are now handling the results in the SDK.

File: src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java (hunk 9 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanBuildStep.java]: SAST_SCA class had been made to handle the SAST+SCA scans execution.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java [hunk 14]
```diff
@@ -28,31 +32,40 @@ public class StaticAnalyzer extends Scanner {
 	private static final String STATIC_ANALYZER = "Static Analyzer"; //$NON-NLS-1$
         
         private boolean m_openSourceOnly;
+        private String m_includeSCAGenerateIRX;
+        private boolean m_additionalOptionsUploadDirect;
+        private boolean m_includeSCAUploadDirect;
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:37]: I think we can simplify this with shorter variable names and we only need 2 new member variables:
```
private boolean m_includeSCA;
private boolean m_uploadDirect;
```
The value of m_includeSCAGenerateIRX is just `m_includeSCA && !m_uploadDirect`. Likewise, the value of m_includeSCAUploadDirect is just `m_includeSCA && m_uploadDirect`. So I don't think we need a third variable here.

File: src/main/java/com/hcl/appscan/jenkins/plugin/actions/ScanResults.java [hunk 15]
```diff
@@ -151,7 +151,7 @@ public File getReport() {
 	}
 	
 	private String getReportName() {
-		String name = (getScanType() + FileUtil.getValidFilename(getName())).replaceAll(" ", ""); //$NON-NLS-1$ //$NON-NLS-2$
+		String name = (FileUtil.getValidFilename(getName())).replaceAll(" ", ""); //$NON-NLS-1$ //$NON-NLS-2$
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/actions/ScanResults.java:154]: sure, Matt.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java (hunk 14 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:37]: Matt, Are you referring that we should use the same name for both the checkbox(under "generateIRX" & "upload Files & folders" scan menthod) in config file? 
1. There is difference in the UI names for the both the checkboxes.
2. Third variable is for the value of the optionalBlock checkbox under the "upload files & folders" scan method. We can remove that optionalBlock as it contains only 1 element into it.

File: src/main/resources/com/hcl/appscan/jenkins/plugin/Messages.properties [hunk 16]
```diff
@@ -16,26 +16,29 @@ error.no.creds=Required. You can add HCL AppScan on Cloud/HCL AppScan 360° cred
...
 error.application.empty.ui = Application list is empty. Please check the credentials for service connection. 
 error.url.validation.ui = Unless this is a private network, please verify your domain.
 error.url.validation = An error occurred while validating the Starting URL: {0}.
 error.sca.AppScan360.ui= Software Composition Analysis is available in AppScan on Cloud only.
 warning.sca.ui= To scan open-source files, use the Software Composition Analysis (SCA) scan type.
-warning.sca= Note: AppScan on Cloud (ASoC) now performs SAST and SCA analysis as separate scans. To execute an open-source only scan, use the Software Composition Analysis (SCA) scan type. The open-source only option will be removed from SAST scans in a future release.
-error.url.dynamic.unsupported= Either your A360 instance does not support dynamic scans or the starting URL is invalid: {0}.
+warning.sca= AppScan on Cloud (ASoC) now performs SAST and SCA analysis as separate scans. To execute an open-source only scan, use the Software Composition Analysis (SCA) scan type.
+warning.include.sca.AppScan360= Only SAST scan execution takes place as SCA is applicable for AppScan on Cloud only.
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/resources/com/hcl/appscan/jenkins/plugin/Messages.properties:39]: I have reviewed the attached statement with Binoy, updating the warning message to  "Only the SAST scan is being executed. SCA is only available with AppScan on Cloud and is not included in this job"

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java (hunk 14 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java]: I have changed the variable name for the additional option under "upload Files & folders" scan method. New variable name would be "m_hasOptionsUploadDirect".
Also, changed the variable type of the "m_includeSCAUploadDirect" to String so that we have a parity between the both checkboxes of "includeSCA".

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java [hunk 17]
```diff
@@ -77,29 +90,84 @@ public String checkScanSpeed(String scanSpeed) {
...
         }
 
-        public boolean isSourceCodeOnly() {
+        public boolean getOpenSourceOnly() {
+            return m_openSourceOnly;
+        }
+        
+        @DataBoundSetter
+        public void setIncludeSCAGenerateIRX(String includeSCAGenerateIRX) {
+            if(getHasOptions() && includeSCAGenerateIRX.equals("true")) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:105]: This can be simplified to:
`return getHashOptions() && Boolean.parseBoolean(includeSCAGenerateIRX);`

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java [hunk 18]
```diff
@@ -77,29 +90,84 @@ public String checkScanSpeed(String scanSpeed) {
...
+        }
+
+        //using this method in the jelly file to determine the checkbox value
+        public boolean getHasOptionsUploadDirect() {
+            return m_hasOptionsUploadDirect;
+        }
+
+        @DataBoundSetter
+        public void setIncludeSCAUploadDirect(String includeSCAUploadDirect) {
+            if(m_hasOptionsUploadDirect && includeSCAUploadDirect.equals("true")) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:139]: Should use `Boolean.parseBoolean(string)`.

**Review by mattmurp — COMMENTED** [review]
There are too many hard coded strings for "true" and "false". When checking for boolean values from strings, you should use `Boolean.parseBoolean(stringValue)` and for setting a string to "true" or "false" you should use `Boolean.toString(booleanValue)`.

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java (hunk 17 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:105]: Okay, sure

File: src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java (hunk 18 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/jenkins/plugin/scanners/StaticAnalyzer.java:139]: Okay, sure.

File: .idea/libraries/Maven__com_hcl_appscan_sdk_1_1_3_SNAPSHOT.xml
Comment (mattmurp) [inline .idea/libraries/Maven__com_hcl_appscan_sdk_1_1_3_SNAPSHOT.xml:1]: Is there a reason you're adding this file? Typically, we don't want files generated by the IDE.

**Review by mattmurp — APPROVED** [review]
The StaticAnalyzer class needs to be cleaned up, but approving in order to get the release done.
